apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/BooleanExpression.kt

```diff
@@ -109,10 +110,11 @@ fun <T : Any> BooleanExpression<T>.evaluate(block: (T) -> Boolean): Boolean {
   }
 }
 
-fun BooleanExpression<BTerm>.evaluate(variables: Set<String>, typename: String?): Boolean {
+fun BooleanExpression<BTerm>.evaluate(variables: Set<String>, typename: String?, adapterContext: AdapterContext, path: String): Boolean {
```

Comment:
I missed that the `path` was part of the fragment identifier. What about replacing `DeferredFragmentIdentifier` with plain strings? And `BLabel` with `BFragmentId`. This way we can have 

```kotlin
fun BooleanExpression<BTerm>.evaluate(variables: Set<String>, typename: String?, fragmentIds: Set<String>): Boolean {
  return evaluate {
    when (it) {
      is BVariable -> variables.contains(it.name)
      is BFragmentId -> fragmentIds.contains(it.id)
      is BPossibleTypes -> it.possibleTypes.contains(typename)
    }
  }
}
```

And `BooleanExpression` remains relatively standalone.

Reply:
> the path was part of the fragment identifier

yes but it's the current path that you can only get while parsing the JSON - that is why we still need to get it from `JsonReader.getPath` and pass it to this method. Unless there's a way around it that I'm missing?

Reply:
Let's continue the discussion in a single thread [below](https://github.com/apollographql/apollo-kotlin/pull/3947#discussion_r832013147)